File: laboneq/compiler/code_generator/wave_index_tracker.py
# Copyright 2022 Zurich Instruments AG
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

from typing import Dict, List, Optional, Tuple, Union


class WaveIndexTracker:
    def __init__(self):
        self._wave_indices: Dict[str, Tuple[int, str]] = {}
        self._next_wave_index: int = 0

    def lookup_index_by_wave_id(self, wave_id: str) -> Optional[int]:
        wave_index_entry = self._wave_indices.get(wave_id)
        return None if wave_index_entry is None else wave_index_entry[0]

    def create_index_for_wave(self, wave_id: str, signal_type: str) -> Optional[int]:
        assert wave_id not in self._wave_indices
        if signal_type == "csv":
            # For CSV store only the signature, do not allocate an index
            self._wave_indices[wave_id] = (-1, signal_type)
            return None

        index = self._next_wave_index
        self._next_wave_index += 1
        self._wave_indices[wave_id] = (index, signal_type)
        return index

    def add_numbered_wave(self, wave_id: str, signal_type: str, index):
        self._wave_indices[wave_id] = (index, signal_type)

    def wave_indices(self) -> Dict[str, List[Union[int, str]]]:
        # Cast entries to list. This will end up in the compiled experiment, and we want
        # invariance under serialization + deserialization, but the JSON decoder
        # produces lists, not tuples.
        return {k: list(v)[:2] for k, v in self._wave_indices.items()}

File: src/rust/seqc-tracker/src/wave_index_tracker.rs
// Copyright 2025 Zurich Instruments AG
// SPDX-License-Identifier: Apache-2.0

use crate::{Error, Result};
use codegenerator::ir::compilation_job::AwgKind;
use indexmap::IndexMap;

pub type WaveIndex = u32;

pub enum SignalType {
    CSV,
    COMPLEX,
    SIGNAL(AwgKind),
}
pub struct WaveIndexTracker {
    pub wave_indices: IndexMap<String, (Option<WaveIndex>, SignalType)>,
    next_wave_index: WaveIndex,
}

impl Default for WaveIndexTracker {
    fn default() -> Self {
        Self::new()
    }
}

impl WaveIndexTracker {
    pub fn new() -> Self {
        Self {
            wave_indices: IndexMap::new(),
            next_wave_index: 0,
        }
    }
    pub fn lookup_index_by_wave_id(&self, wave_id: &str) -> Option<Option<WaveIndex>> {
        let entry = self.wave_indices.get(wave_id);
        entry.map(|entry| entry.0)
    }

    pub fn create_index_for_wave<S1: Into<String>>(
        &mut self,
        wave_id: S1,
        signal_type: SignalType,
    ) -> Result<Option<WaveIndex>> {
        let wave_id: String = wave_id.into();
        if self.wave_indices.contains_key(&wave_id) {
            return Err(Error::new("Wave ID already exists"));
        }
        let index = if matches!(signal_type, SignalType::CSV) {
            // For CSV store only the signature, do not allocate an index
            None
        } else {
            let index = self.next_wave_index;
            self.next_wave_index += 1;
            Some(index)
        };
        self.wave_indices.insert(wave_id, (index, signal_type));
        Ok(index)
    }

    pub fn add_numbered_wave<S1: Into<String>>(
        &mut self,
        wave_id: S1,
        signal_type: SignalType,
        index: WaveIndex,
    ) {
        self.wave_indices
            .insert(wave_id.into(), (Some(index), signal_type));
    }

    pub fn finish(self) -> IndexMap<String, (Option<WaveIndex>, SignalType)> {
        self.wave_indices
    }
}
